contracts/secretswap_router/src/contract.rs

```diff
@@ -0,0 +1,378 @@
+                        )?)
+                    } else {
+                        // not last hop
+                        // 1. set expected_return to None because we don't care about slippage mid-route
+                        // 2. set the recepient of the swap to be this contract (the router)
+                        msgs.push(snip20::send_msg(
+                            next_hop.pair_address,
+                            amount,
+                            Some(to_binary(&Snip20Swap::Swap {
+                                expected_return: None,
+                                to: Some(env.contract.address.clone()),
+                            })?),
+                            None,
+                            256,
+                            next_hop.from_token.code_hash.unwrap(), // tested before first hop
+                            next_hop.from_token.address.unwrap(),   // tested before first hop
+                        )?)
+                    }
+
+                    let remaining_hops: Vec<Hop> = remaining_route.hops[1..].to_vec();
+
+                    store_route_state(
+                        &mut deps.storage,
+                        &RouteState {
+                            is_done,
+                            remaining_route: Route {
+                                hops: remaining_hops,
+                                to: remaining_route.to.clone(),
+                            },
+                        },
+                    )?;
+
+                    Ok(HandleResponse {
+                        messages: msgs,
+                        log: vec![],
+                        data: None,
+                    })
+                }
+                None => Err(StdError::generic_err("cannot find route")),
+            }
+        }
+        HandleMsg::FinalizeRoute {} => match read_route_state(&deps.storage)? {
+            Some(RouteState {
+                is_done,
+                remaining_route,
+            }) => {
+                // this function is called only by the route creation function
+                // it is intended to always make sure that the route was completed successfully
+                // otherwise - revert the transaction
+
+                if env.contract.address != env.message.sender {
+                    return Err(StdError::unauthorized());
+                }
+                if !is_done {
+                    return Err(StdError::generic_err("cannot finalize: route is not done"));
+                }
+                if remaining_route.hops.len() != 0 {
+                    return Err(StdError::generic_err(
+                        "cannot finalize: route still has hops",
+                    ));
+                }
+
+                delete_route_state(&mut deps.storage);
+
+                Ok(HandleResponse::default())
+            }
+            None => Err(StdError::generic_err("no route to finalize")),
+        },
+        HandleMsg::RegisterTokens { tokens } => {
```

Comment:
Why are you accepting a Token structure here that can be missing parameters you need? Just define a new structure without the optionals